### ISSUE TITLE
Better maintainable whitelists

### ISSUE THREAD

**Issue by nebulon42** [gravitystorm/openstreetmap-carto#2452]
I decided to open a separate issue for that. It is related to #2444 and an unmaintainable white list of shops was the reason that #2415 got merged.

So we need better maintainable whitelists. One idea would be to put them into the database. One then would only have to maintain the content of the tables and the SQL would be automatically adjusted. This of course open up questions of how the rendering database gets updated with the white lists.

Basic proof of concept:

````
CREATE TABLE barrier_whitelist (feature TEXT PRIMARY KEY);

INSERT INTO barrier_whitelist VALUES
  ('chain'),
  ('city_wall'),
  ('embankment'),
  ('ditch'),
  ('fence'),
  ('guard_rail'),
  ('handrail'),
  ('hedge'),
  ('kerb'),
  ('retaining_wall'),
  ('wall');
````
````
        (SELECT
            way, COALESCE(historic, barrier) AS feature
          FROM (SELECT way,
            ('barrier_' || (CASE WHEN barrier IN (SELECT feature FROM barrier_whitelist) THEN barrier ELSE NULL END)) AS barrier,
            ('historic_' || (CASE WHEN historic = 'citywalls' THEN historic ELSE NULL END)) AS historic
            FROM planet_osm_line
            WHERE barrier IN (SELECT feature FROM barrier_whitelist)
              OR historic = 'citywalls'
              AND (waterway IS NULL OR waterway NOT IN ('river', 'canal', 'derelict_canal', 'stream', 'drain', 'ditch', 'wadi'))
          ) AS features
        ) AS line_barriers
````

In pgAdmin it works well, with Mapnik I have the problem that some initializing is done and I run into:
````
Postgis Plugin: ERROR:  column "way" does not exist
LINE 1: SELECT ST_SRID("way") AS srid FROM barrier_whitelist WHERE "...
                       ^
in executeQuery Full sql was: 'SELECT ST_SRID("way") AS srid FROM barrier_whitelist WHERE "way" IS NOT NULL LIMIT 1;'
  encountered during parsing of layer 'line-barriers' in Layer
````
I'm not sure if we can get around this.

Other ideas?

**Comment by imagico** [comment]
> I'm not sure if we can get around this.

According to mapnik docs:

https://github.com/mapnik/mapnik/wiki/PostGIS

specifying an srid parameter for the postgis data source of the layer _will avoid an extra database query for knowing the srid of the table_.

If that does not help - well i guess you could always add a bogus geometry field to your table...

**Comment by pnorman** [comment]
> Also specifying `geometry_table` might be intended for such cases - although the docs are not quite clear here.

Yes, if you specify the table name it will know what to query.

**Comment by pnorman** [comment]
Adding external SQL makes it more difficult for people to get started. It's been my experience that this can actually be a significant barrier to new people.

**Comment by nebulon42** [comment]
> Adding external SQL makes it more difficult for people to get started. It's been my experience that this can actually be a significant barrier to new people.

Yes, that is an important point. I'm wondering if there are better ways or if we have to live with some addition of complexity to move forward.

**Comment by kocio-pl** [comment]
Maybe using variables and/or storing lists in separate file(s)? That would make project.mml more readable.

**Comment by kocio-pl** [comment]
@nebulon42 Does CartoCSS allow using variables so they could contain such lists, so instead of:

```SQL
('barrier_' || (CASE WHEN barrier IN (SELECT feature FROM barrier_whitelist) THEN barrier ELSE NULL END)) AS barrier,
```

we could use just:

```SQL
('barrier_' || (CASE WHEN barrier IN @barrier_whitelist THEN barrier ELSE NULL END)) AS barrier,
```

?

**Comment by matthijsmelissen** [comment]
CartoCSS itself does not, but fortunately YAML supports this (see YAML anchors). I think we should replace all repeated data inthe yaml by anchors.

**Comment by kocio-pl** [comment]
I thought that CartoCSS preprocesses all the files. But particular solution is not important - that would be great to have such lists! My current biggest pain when editing project.mml is that... I don't have a 50" monitor and I have to scroll all of the time. :smile:

**Comment by matthijsmelissen** [comment]
Could you have a look if it works? YAML achors don’t look to hard to use.

**Comment by kocio-pl** [comment]
Sure, I'll try.

**Comment by kocio-pl** [comment]
Unfortunately, using anchors is limited and the name of this feature reflects the difference: 

https://stackoverflow.com/questions/4150782/use-yaml-with-variables#18877077

So one way would be to process YAML in addition to MSS to act like all the other variables, as asked, or to chop SQL queries into fields like (this might be not proper YAML code, it's just a visual example):

```yaml
  list: &barrier_whitelist
    - gate
    - block 
  start_of_the query: ('barrier_' || (CASE WHEN barrier IN (SELECT feature FROM 
  whitelist: *barrier_whitelist
  rest_of_the query: ) THEN barrier ELSE NULL END)) AS barrier,
```

but that looks ugly and while it would be proper YAML, the SQL query would be hard to read/parse and should be concatenated anyway (and that is a simple case with just one whitelist - what to do with more of them?).

I think proposed solution is quite straightforward preprocessing: straight substituting strings with another strings in YAML, no matter if it contains SQL or anything else.

@nebulon42 What do you think about it now?

**Comment by matthijsmelissen** [comment]
@kocio-pl Thanks for investigating, didnt know it was that limited!

**Comment by dch0ph** [comment]
#5169 introduced an approach for white-listing based on taginfo usage. I feel this covers the most of the problem cases (except the long list of shops with specific symbols).

Most other cases involve relative short numbers of specific tags, where it is helpful to be able to see the values.

We do have the option of using of SQL function to avoid repetitions, e.g. a function to determine whether a `surface` value is paved or unpaved.

Overall, I think this issue has largely evaporated.

**Comment by imagico** [comment]
Closing, as we have introduced a solution using database tables.
